ports/esp32s2/common-hal/socketpool/Socket.c

```diff
@@ -119,16 +124,74 @@ mp_uint_t common_hal_socketpool_socket_recv_into(socketpool_socket_obj_t* self,
     return received;
 }
 
+mp_uint_t common_hal_socketpool_socket_sendto(socketpool_socket_obj_t* self,
+    const char* host, size_t hostlen, uint8_t port, const uint8_t* buf, mp_uint_t len) {
+
+    // Get the IP address string
+    const struct addrinfo hints = {
+        .ai_family = AF_INET,
+        .ai_socktype = SOCK_STREAM,
+    };
+    struct addrinfo *result;
+    int error = lwip_getaddrinfo(host, NULL, &hints, &result);
+    if (error != 0 || result == NULL) {
+        return 0;
+    }
+    #pragma GCC diagnostic push
+    #pragma GCC diagnostic ignored "-Wcast-align"
+    struct in_addr *addr = &((struct sockaddr_in *)result->ai_addr)->sin_addr;
+    #pragma GCC diagnostic pop
+    char ip_str[IP4ADDR_STRLEN_MAX];
+    inet_ntoa_r(*addr, ip_str, IP4ADDR_STRLEN_MAX);
+    freeaddrinfo(result);
+
+    // Set parameters
+    struct sockaddr_in dest_addr;
+    dest_addr.sin_addr.s_addr = inet_addr((const char *)ip_str);
+    dest_addr.sin_family = AF_INET;
+    dest_addr.sin_port = htons(port);
+
+    int bytes_sent = lwip_sendto(self->num, buf, len, 0, (struct sockaddr *)&dest_addr, sizeof(dest_addr));
+    if (bytes_sent < 0) {
+        mp_raise_BrokenPipeError();
+        return 0;
+    }
+    return bytes_sent;
+}
+
+mp_uint_t common_hal_socketpool_socket_recvfrom_into(socketpool_socket_obj_t* self,
+    uint8_t* buf, mp_uint_t len, uint8_t* ip, uint *port) {
+
+    struct sockaddr_in source_addr;
+    socklen_t socklen = sizeof(source_addr);
+    int bytes_received = lwip_recvfrom(self->num, buf, len - 1, 0, (struct sockaddr *)&source_addr, &socklen);
+
+    memcpy((void *)ip, (void*)&source_addr.sin_addr.s_addr, sizeof source_addr.sin_addr.s_addr);
+    *port = source_addr.sin_port;
+
+    if (bytes_received < 0) {
+        mp_raise_BrokenPipeError();
+        return 0;
+    } else {
+        buf[bytes_received] = 0; // Null-terminate whatever we received
+        return bytes_received;
+    }
+}
+
 void common_hal_socketpool_socket_close(socketpool_socket_obj_t* self) {
     self->connected = false;
     if (self->tcp != NULL) {
         esp_tls_conn_destroy(self->tcp);
         self->tcp = NULL;
     }
+    if (self->num >= 0) {
+        lwip_shutdown(self->num, 0);
+        lwip_close(self->num);
+    }
```

Comment:
Should self->num be assigned -1 here so that the closed socket can be recognized?  Otherwise, wont a subsequent call risk doing lwip API calls on some subsequently created socket?

Reply:
seems reasonable